### PR TITLE
JENKINS-47114# Upgrade plugin-pom and jenkins-test-harness

### PR DESCRIPTION
**Pull request by vivek**
Fixes running unit tests from inside IDEA.

# Description

See [JENKINS-47114](https://issues.jenkins-ci.org/browse/JENKINS-47114).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by vivek** [comment]
Multiple test failures due to probably issue with build agent. Running it again.
 
```
getOrganizationsAndRepositories(io.jenkins.blueocean.blueocean_github_pipeline.GithubApiTest)  Time elapsed: 1.043 sec  <<< ERROR!
com.github.tomakehurst.wiremock.common.FatalStartupException: java.lang.RuntimeException: java.lang.IllegalStateException: Insufficient threads: max=10 < needed(acceptors=4 + selectors=8 + request=1)
	at com.github.tomakehurst.wiremock.WireMockServer.start(WireMockServer.java:146)
	at com.github.tomakehurst.wiremock.junit.WireMockRule$1.evaluate(WireMockRule.java:68)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:533)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: java.lang.IllegalStateException: Insufficient threads: max=10 < needed(acceptors=4 + selectors=8 + request=1)
	at com.github.tomakehurst.wiremock.jetty9.JettyHttpServer.start(JettyHttpServer.java:132)
	at com.github.tomakehurst.wiremock.WireMockServer.start(WireMockServer.java:144)
	... 6 more
Caused by: java.lang.IllegalStateException: Insufficient threads: max=10 < needed(acceptors=4 + selectors=8 + request=1)
	at org.eclipse.jetty.server.Server.doStart(Server.java:414)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at com.github.tomakehurst.wiremock.jetty9.JettyHttpServer.start(JettyHttpServer.java:130)
	... 7 more
```

**Comment by michaelneale** [comment]
@vivek build failing due to code coverage violations - any idea why when nothing there has changed?

**Comment by vivek** [comment]
@michaelneale no, code coverage is fine. Scroll up to see the error, its `Caused by: java.lang.IllegalStateException: Insufficient threads: max=10 < needed(acceptors=4 + selectors=8 + request=1)
`

The failure is legit, first I thought its build agent issue, but running locally I could reproduce as well. Looking in to it.

**Comment by vivek** [comment]
@michaelneale looks like caused by https://github.com/jenkinsci/jenkins-test-harness/pull/69/files. jenkins-test-harness limits acceptor and selector to hard coded 1, wire mock uses default jetty config where it requires 4 and 8 respectively. Checking if there is way to configure jetty in wire mock.

**Comment by michaelneale** [comment]
@vivek ok

**Comment by michaelneale** [comment]
@vivek well I can certainly run this test individually locally, so that is good... now if we can get the thread thing sorted - should get this in.

**Comment by michaelneale** [comment]
@vivek if you get this passing (and will need to fix the JWT test, which should be fixed on master shortly) then get it in. Locally worked well.

**Comment by vivek** [comment]
@michaelneale its bit complicated that error (thread one) I could fix but there is mess around jetty version in use by jenkins-test-harness(9.4.5) and wiremock(9.2.22). There is API incompatibility between them and we are picking one of the higher version (test-harness) but then that causes runtime failure:

```
java.lang.NoSuchMethodError: org.eclipse.jetty.util.ssl.SslContextFactory.selectCipherSuites([Ljava/lang/String;[Ljava/lang/String;)[Ljava/lang/String;
```

Well picking one vs another version won't work because both wire mock and test-harness use these version specific API. So fix is really, either test-harness downgrades jetty version to 9.2.22 or wire mock upgrades jetty to 9.4.5.

I have sent details to @jglick and trying to figure best course of action/workaround(if any).

**Comment by michaelneale** [comment]
@vivek if no quick solution could you post to jenkins-ux mailing list with instructions on how to run individual tests so others can reference it?

**Comment by vivek** [comment]
@michaelneale sure will do

**Comment by michaelneale** [comment]
getting closer. A few errors now seem to be related to eclipse/jetty I guess: 

```
createPersonalSSHKey – io.jenkins.blueocean.ssh.UserSSHKeyTest
a few seconds
anonymousUserToken – io.jenkins.blueocean.auth.jwt.impl.JwtAuthenticationServiceImplTest
a few seconds
getToken – io.jenkins.blueocean.auth.jwt.impl.JwtAuthenticationServiceImplTest
a few seconds
testOrg – io.jenkins.blueocean.service.embedded.BlueOceanWebURLBuilderTest
a few seconds
test_freestyle – io.jenkins.blueocean.service.embedded.BlueOceanWebURLBuilderTest
a few seconds
testCustomerUrlMapper – io.jenkins.blueocean.service.embedded.BlueOceanWebURLBuilderTest
a few seconds
```

eg:

```
org/eclipse/jetty/security/UserStore
Stacktrace
java.lang.NoClassDefFoundError: org/eclipse/jetty/security/UserStore
	at org.jvnet.hudson.test.JenkinsRule.configureUserRealm(JenkinsRule.java:685)
	at org.jvnet.hudson.test.JenkinsRule.createWebServer(JenkinsRule.java:659)
```

**Comment by vivek** [comment]
Almost there, rebased with master to pull ssh user key test fix and other latest stuff.

**Comment by michaelneale** [comment]
:bee: works lovely @vivek

File: pom.xml [hunk 1]
```diff
@@ -790,57 +816,74 @@
...
                     <artifactId>asm</artifactId>
                 </exclusion>
             </exclusions>
         </dependency>
+        <!--
+            wiremock jetty, depends on 9.2.22.v20170606, jenkins-test-harness on jetty 9.4.5.
+
+            jetty-util 9.4.5 breaks backward compatibility, where
+
+            SslContextFactory.selectCipherSuites([Ljava/lang/String;[Ljava/lang/String;)[Ljava/lang/String;
```
Comment (jtnord) [inline pom.xml]: https://github.com/tomakehurst/wiremock/issues/551#issuecomment-268647572

use wiremock-standalone that shades jetty...  did you try this?

Comment (vivek) [reply pom.xml]: @jtnord thanks for the pointer, no I haven't. Happy to try as it will clean all that work around.

**Pull request merged**